Step5 코멘트 반영 후 pr 재요청드립니다.

hyunhyun opened the pull request:
안녕하세요~ step 5 로또 수동 구매 리뷰 요청드립니다~
이전 리뷰에서 코멘트 해주신 내용들도 반영하였습니다.
감사합니다!

hyunhyun committed:
docs(README.md) : step3 기능요구사항 정리

hyunhyun committed:
feat : 구입금액에 따른 로또 여러건 생성

- LottoTest : 금액 입력 로또 개수 확인 테스트코드
- LottoMachine : 로또 여러건 생성, 구매내역 출력
- Lottos : 로또 여러장 저장 컬렉션
- Lotto : 로또 내용 출력

hyunhyun committed:
feat : 당첨 개수 구하기, 총 당첨금액 구하기, 당첨결과 출력

Lotto : 당첨로또 비교 당첨개수 구하기
Lottos : 총 당첨금액 구하기, 등수별 개수 구하기
ResultView : 등수별 당첨 통계 출력
Rank : 당첨 순위 정보 도메인
LottoTest : 당첨순위 확인, 당첨금액 확인 테스트코드 작성

hyunhyun committed:
refactor : LottoMachine 삭제

hyunhyun committed:
refactor : Rank enum으로 변경

hyunhyun committed:
test : 4등 당첨 확인 테스트 추가

hyunhyun committed:
docs : step4 로또2등 보너스 기능요구사항 작성

hyunhyun committed:
format : merge오류 지우기

hyunhyun committed:
fix : 당첨통계 출력 오류 수정

hyunhyun committed:
refactor : 로또번호 원시값 포장 (int -> LottoNo)

hyunhyun committed:
test : LottoNo 유효성 테스트 추가

hyunhyun committed:
refactor : 수익률계산 반환값 타입 변, ]

WinRank : 수익률계산 반환값 타입 변경
Rank : 순위 반환 코드 리팩토링

hyunhyun committed:
feat : step5 로또 수동 구매 추가

jaeyeonling commented on src/main/java/lotto/LottoPlay.java:
```diff
@@ -1,21 +1,47 @@
 import lotto.ui.ResultView;
 
 public class LottoPlay {
+    static int moneyInput;
+    static int manualCount;
+    static int autoCount;
+    static Lottos lottos;
+    static Lotto winningLottoInput;
+    static LottoNo bonusNumber;
```
해당 값들을 static으로 선언하지 않고 구현해 볼 수 있을 것 같은데요.
static으로 구현하는 것과 인스턴스 변수로 구현하는 것은 어떠한 차이가 있을지도 고민해보아요.

hyunhyun replied:
main 메소드가 너무 길어져서 분리하는데만 신경을 쓰단보니 static으로 계속 저장될 필요가 없는 변수들까지 static으로 처리가 되서 메모리 낭비가 발생했네요. 메소드를 더 분리해서 파라미터로 필요한 값 넘기도록 수정했습니다.

jaeyeonling commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,54 @@
+        lottoNumbers.add(lottoNo);
+    }
+
+    public boolean checkBonusMatch(LottoNo bonusNumber) {
```
매개변수로 넘어오는 값이 bonusNumber가 아니어도 동작할 것 같은데요.
그렇다면 메서드의 이름이나 매개변수의 이름이 변경돼야 하지 않을까요?

hyunhyun replied:
소스에서 로또랑 로또 번호 일치하는 개수 구할때도 해당 메소드를 부를 수 있더라구요. containsLottoNo로 변경하겠습니다.

jaeyeonling commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,54 @@
+        return lottoNumbers;
+    }
+
+    public int checkMatchCount(Lotto checkLotto) {
```
check-라는 이름은 일반적으로 boolean 값을 응답할 때 사용하는데요.
해당 메서드의 기능을 나타낼 수 있는 이름이 없을지 고민해보아요.

jaeyeonling commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,54 @@
+    public static Lotto createRandomLotto(List<Integer> lottoNumbers) {
+        Lotto lotto = new Lotto();
+        for (Integer lottoNumber : lottoNumbers) {
+            lotto.getLottoNumbers().add(new LottoNo(lottoNumber));
```
getter를 사용하지 않고 메시지를 던지는 방식으로 구현해보면 좋겠네요~

hyunhyun replied:
아래 addLottoNumber 함수를 활용 하도록 변경했습니다. 그리고 createRandomLotto로 명을 지어서 랜덤으로 생성된 값으로 로또 사용할때만 사용했었는데 일반 생성자로 만들어서 테스트코드에서 List<Integer> 타입으로도 로또를 생성할 수 있도록 하였습니다.

jaeyeonling commented on src/main/java/lotto/domain/LottoNo.java:
```diff
@@ -0,0 +1,59 @@
+        } catch (NumberFormatException numberFormatException) {
+            throw new IllegalArgumentException("로또볼은 숫자입니다.");
+        }
+        this.lottoNumber = validateNumberRange(checkLottoNoInput);
```
주생성자를 정하여 인스턴스 변수는 주생성자를 통해서만 초기화하도록 수정해보면 좋겠어요~

hyunhyun replied:
넵 감사합니다~

jaeyeonling commented on src/main/java/lotto/domain/LottoNo.java:
```diff
@@ -0,0 +1,59 @@
+        return String.valueOf(lottoNumber);
+    }
+
+    public LottoNo validateBonus(Lotto winningLotto) {
```
```
class A {
    상수(static final) 또는 클래스 변수
    인스턴스 변수
    생성자
    팩토리 메소드
    메소드
    기본 메소드 (equals, hashCode, toString)
}
```

클래스 구현 순서는 위와 같습니다!

jaeyeonling commented on src/main/java/lotto/domain/WinRanks.java:
```diff
@@ -28,23 +29,26 @@ public int winningPrice(Lotto winningLotto, Lottos lottos, int bonusNumber) {
-    private void addRankCount(Map<Rank, Integer> winTotals, int checkMatchCount, boolean bonusMatch) {
-        Rank key = Rank.matchedRank(checkMatchCount, bonusMatch);
+    private void addRankCount(int checkMatchCount, boolean bonusMatch) {
+        Optional<Rank> rank = Rank.matchedRank(checkMatchCount, bonusMatch);
```
rank가 없을 경우 어떻게 동작하는게 맞을까요?
예외처리를 한다면 어디서 하는게 좋을지도 고민해보아요.

hyunhyun replied:
stream을 잘 활용하지 못했던 거 같습니다. 수정하였습니다!

jaeyeonling commented on src/main/java/lotto/domain/WinRanks.java:
```diff
@@ -28,23 +29,26 @@ public int winningPrice(Lotto winningLotto, Lottos lottos, int bonusNumber) {
+    private void addRankCount(int checkMatchCount, boolean bonusMatch) {
+        Optional<Rank> rank = Rank.matchedRank(checkMatchCount, bonusMatch);
+        Rank key = null;
+        if(rank.isPresent()){
```
```suggestion
        if (rank.isPresent()) {
```

컨벤션을 확인해주세요~

hyunhyun replied:
넵! 개발 후에 reformat하는거 자꾸 놓치네요..

jaeyeonling commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,65 @@
+
+    static Lotto lotto;
+
+    static class TestLotto extends Lotto {
```
TestLotto를 따로 만들어주신 이유가 있을까요?

hyunhyun replied:
생성자가 List<Integer> 가 별도로 있었는데 테스트코드에서만 사용하고 있어서 지웠더니 소스코드에 String[] 타입 생성자 밖에 없었는데, 테스트 코드 작성이 너무 불편하여 테스트 내부 클래스를 만들었었는데
[코멘트](https://github.com/next-step/java-lotto-pro/pull/626#discussion_r884139291) 에서 답변해주신거 메소드를 생성자변경하도록해서 테스트코드에서 이를 활용하였습니다.

jaeyeonling commented on src/main/java/lotto/domain/LottoConst.java:
```diff
@@ -0,0 +1,8 @@
+package lotto.domain;
+
+public class LottoConst {
```
상수들을 따로 분리해주신 이유가 있을까요?

hyunhyun replied:
Lotto와 관련된 값들이 중복되어서 상수로 처리하려고 빼었는데, 일급컬렉션인 Lotto에 다른 변수는 없어야되는줄 알고 별도로 분리하였ㅡ는데 일급컬렉션에는 다른 멤버변수만 없으면 되고 클래스변수는 괜찮은 거였네요. Lotto안으로 변경하였습니다.

jaeyeonling reviewed (changes requested):
마지막 단계 미션을 잘 구현해주셨어요 👍 
추가적인 코멘트 남겨두었으니 확인 부탁드려요!!

hyunhyun committed:
refactor : 불필요 static 정리, 당첨순위
들 계산 수정, 메소드 재사용

Lotto : 메소드 List<Integer> 타입 생성자로 수정, 상수 추가
LottoPlay : 불필요 static 변수 삭제
winRanksTest, LottoTest :TestLotto 내부클래스 삭제
LottoNo : 주 생성자 하나로 수정
Rank : matechedRank 삭제
WinRanks : 당첨순위들 계산 리팩토링

hyunhyun committed:
refactor : LottoConst 삭제

jaeyeonling commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,57 @@
+        lottoNumbers.add(lottoNo);
+    }
+
+    public boolean containsLottoNo(LottoNo bonusNumber) {
```
아직도 bonusNumber라는 이름이 남아있네요 ㅎㅎ

jaeyeonling reviewed (approved):
미션을 잘 진행해주셨어요!!
수 많은 코멘트 반영하시느라 정말 고생 많으셨습니다~
해당 요청은 머지하도록 하겠습니다 마지막까지 화이팅입니다 😄
